Standards/scs-0213-v1-k8s-nodes-anti-affinity.md

```diff
@@ -0,0 +1,123 @@
+they would be preferred if possible.
+
+These features allow efficient resource usage (e.g. by scheduling workloads to evenly
+distribute across nodes) and provide fault-tolerance and therefore high availability.
+But they also require more work during the setup of a service architecture, since nodes
+and pods need to be labelled and described consistently.
+
+In the case of SCS, affinity of the workloads themselves is not relevant, since this
+feature is mostly used by the customers of the providers.
+Instead, the expected standard should enable the Kubernetes cluster to handle Anti Affinity
+rules with a real physical separation as well as distributing the control plane over
+multiple machines in order to provide fault-tolerance during system outages.
+If the control plane survives an outage, a Kubernetes cluster can recover later on.
+
+## Design considerations
+
+SCS plans to require a Hard Anti Affinity and/or Redundancy for the control plane and
+a Soft Anti Affinity for workers. This means, that Anti Affinity would be required for
+the control planes and their pods and only optional (but encouraged) for workers.
+
+In order to achieve the goals for these components, meaning availability and fault tolerance
+for the control plane, in order to achieve an outage resistant cluster, and the availability
+promise given with Anti Affinity for pods on the worker nodes, a separation of nodes
+on the hardware or hypervisor level would need to be achieved. Since running multiple
+hypervisors on the same machine is in most cases not feasible and this approach
+wouldn't change availability for outages of individual machines, this should not be recommended.
+
+For the control plane, a reference to the "Kubernetes High Availability" [1]
+can be useful, since it provides two ways to set up a highly available cluster.
+Both approaches are very similar. The "Stacked Control Plane" requires less infrastructure,
+but also runs the risk of failed coupling, where if one node fails, the redundancy could be
+compromised due to the loss of a complete control plane instance.
+The "External ETCD" solves this problem, but also requires double the infrastructure, due
+to the externally incorporated etcd clusters.
+
+This also shows, that the wording "anti affinity" as used with Kubernetes pods is probably
+slightly misleading in the context of a Kubernetes control plane. It may consist of multiple
+pods with individual tasks, but distributing them over different nodes through Anti Affinity
+would probably still cascade the whole control plane into a complete failure, if one of
+the used nodes goes down. It could be possible to replicate specific important pods and
+assign them to different nodes, but at this point, a redundant setup like presented in [1] could be used.
+So Anti Affinity in this context probably means more like distribution over multiple
+physical machines, which needs to be planned beforehand on the machine/server level.
```

Comment:
Agree, IMO, the k8s control plane (CP) components, and the k8s "anti-affinity" feature should not be mentioned together. 

AFAIK CP components are distributed over the k8s nodes via a mechanism called  `static pod manifests` (in the case when `kubeadm` is used for k8s cluster bootstraping).   `kubeadm` generates static pod manifests for CP components to the `/etc/kubernetes/manifests` directory on each node. Then the `kubelet` watches this directory for Pods to create on startup, see [docs](https://kubernetes.io/docs/reference/setup-tools/kubeadm/implementation-details/#generate-static-pod-manifests-for-control-plane-components).

Reply:
Since this document is only a Decision Record, I'm fine with mentioning them here together (just to address the Issue description), but in the final standard, I would choose another description/name for everything, like "Distribution" or just "High Availability", since IMO this is what this whole idea comes from.

Reply:
It makes sense, thanks for clarifying